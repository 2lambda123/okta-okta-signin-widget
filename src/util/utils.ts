/* eslint-disable max-statements */
/*!
 * Copyright (c) 2015-2016, Okta, Inc. and/or its affiliates. All rights reserved.
 * The Okta software accompanied by this notice is provided pursuant to the Apache License, Version 2.0 (the "License.")
 *
 * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0.
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *
 * See the License for the specific language governing permissions and limitations under the License.
 */

/* eslint complexity: [2, 13], max-depth: [2, 4] */

import _cloneDeep from 'clone-deep';
import deepMerge from 'deepmerge';
import deepEqual from 'fast-deep-equal';

export function cloneDeep<T>(v: T): T {
  return _cloneDeep(v);
}

export function merge<T>(target: Partial<T>, source: Partial<T> | Record<string, unknown>): T {
  return deepMerge(target, source as Partial<T>);
}

export function debounce(func, wait = 0) {
  let timer;
  return (...args) => {
    clearTimeout(timer);
    timer = setTimeout(() => {
      func(...args);
    }, wait);
  };
}

export function set(obj, path, v) {
  const keys = path
    .split(/\.|\[/)
    .map((k) => {
      return k[k.length-1] === ']' ? parseInt(k.substring(0, k.length-1)) : k;
    });
  let target = obj;
  for (let i = 0 ; i < keys.length ; i++) {
    const key = keys[i];
    const nextKey = i < keys.length-1 ? keys[i+1] : undefined;
    if (nextKey) {
      if (target[key] === undefined) {
        if (typeof nextKey === 'number') {
          target[key] = [];
        } else {
          target[key] = {};
        }
      }
      target = target[key];
    } else {
      target[key] = v;
    }
  }
  return obj;
}

export function groupBy<T>(arr: T[], iteratee: (T) => string): Record<string, T[]> {
  const obj = {};
  arr.forEach((v) => {
    const group = iteratee(v);
    if (!obj[group]) {
      obj[group] = [];
    }
    obj[group].push(v);
  });
  return obj;
}

export function flow(...transformFns) {
  return (v) => {
    return transformFns.flat(Infinity).reduce((res, transformFn) => {
      return transformFn(res);
    }, v);
  };
}

export function isObject(v: unknown) {
  return typeof v === 'object' && v !== null && !Array.isArray(v);
}

export function compare(a: unknown, b: unknown) {
  if (a < b) {
    return -1;
  } else if (a > b) {
    return 1;
  } else {
    return 0;
  }
}

export function pick(obj: Record<string, unknown>, keys: string[]) {
  return Object.fromEntries(
    keys
      .filter((k) => obj[k] !== undefined)
      .map((k) => ([k, obj[k]]))
  );
}

export function omit(obj: Record<string, unknown>, keys: string[]) {
  return Object.fromEntries(
    Object.keys(obj)
      .filter((k) => !keys.includes(k))
      .map((k) => ([k, obj[k]]))
  );
}

export function isEmpty(v: unknown) {
  if (isObject(v)) {
    return Object.keys(v).length == 0;
  }
  if (Array.isArray(v) || typeof v === 'string') {
    return v?.length === 0;
  }
  return !v;
}

export function union(...arrs) {
  return arrs.reduce((res, arr) => {
    return [...new Set([...res, ...arr])];
  }, []);
}

export function isEqual(a: unknown, b: unknown) {
  return deepEqual(a, b);
}
