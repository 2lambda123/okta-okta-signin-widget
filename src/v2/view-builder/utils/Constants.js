export const SHOW_RESEND_TIMEOUT = 30000;
export const WARNING_TIMEOUT = 30000;
export const UNIVERSAL_LINK_POST_DELAY = 500;