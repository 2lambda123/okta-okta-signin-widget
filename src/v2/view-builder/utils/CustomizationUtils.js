import Enums from 'util/Enums';

/**
 * @param iconClass icon class to override
 * @param logoUrl URL where the logo can be retrieved
 */
export function overrideIconForClass(iconClass, logoUrl) {
  const template = `
    #okta-sign-in.auth-container .${iconClass} {
      background-image: url('${logoUrl}')
    }
  `;
  const main = document.getElementById(Enums.WIDGET_CONTAINER_ID);
  const style = document.createElement('style');
  style.id = 'okta-sign-in-config-logos'; // TODO: make a const
  style.type = 'text/css';
  style.appendChild(document.createTextNode(template));
  main.appendChild(style);
}