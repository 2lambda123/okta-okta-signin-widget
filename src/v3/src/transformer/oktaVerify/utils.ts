/*
 * Copyright (c) 2022-present, Okta, Inc. and/or its affiliates. All rights reserved.
 * The Okta software accompanied by this notice is provided pursuant to the Apache License, Version 2.0 (the "License.")
 *
 * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0.
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *
 * See the License for the specific language governing permissions and limitations under the License.
 */

import { IdxAuthenticator } from '@okta/okta-auth-js';
import set from 'lodash/set';
import { loc } from 'okta';
import {
  DescriptionElement,
  ListElement,
  QRCodeElement,
  StepperNavButtonConfigAttrs,
  StepperNavButtonConfigDirection,
  UISchemaElement,
} from 'src/types';

export const appendDescriptionElements = (
  elements: UISchemaElement[],
  authenticator?: IdxAuthenticator,
): void => {
  if (!authenticator?.contextualData) {
    return;
  }

  const { contextualData } = authenticator;

  if (contextualData.qrcode) {
    const { href } = contextualData.qrcode;
    const { displayName } = authenticator;
    elements.push({
      type: 'List',
      options: {
        items: [
          loc('oie.enroll.okta_verify.qrcode.step1', 'login'),
          loc('oie.enroll.okta_verify.qrcode.step2', 'login'),
          loc('oie.enroll.okta_verify.qrcode.step3', 'login'),
        ],
        type: 'ordered',
      },
    } as ListElement);

    elements.push({
      type: 'QRCode',
      options: {
        label: displayName,
        data: href,
      },
    } as QRCodeElement);
    return;
  }

  // @ts-ignore OKTA-496373 - missing props from interface
  const { email } = contextualData;
  if (email) {
    elements.push({
      type: 'Description',
      options: {
        // TODO: revist this to use oie i18n string (EnrollChannelPollDescriptionView.js)
        content: loc('next.enroll.okta_verify.email.info', 'login', [email]),
      },
    } as DescriptionElement);
    return;
  }

  // @ts-ignore OKTA-496373 - missing props from interface
  const { phoneNumber } = contextualData;
  if (phoneNumber) {
    elements.push({
      type: 'Description',
      options: {
        // TODO: revist this to use oie i18n string (EnrollChannelPollDescriptionView.js)
        content: loc('next.enroll.okta_verify.sms.info', 'login', [phoneNumber]),
      },
    } as DescriptionElement);
  }
};

export const createNavButtonConfig = (hasQrCode: boolean): {
  navButtonsConfig: Record<StepperNavButtonConfigDirection, StepperNavButtonConfigAttrs>,
} => {
  const stepperNavBtnConfig = {
    navButtonsConfig: {
      next: {
        variant: 'secondary',
        label: hasQrCode
          ? 'enroll.totp.cannotScan'
          : 'next.enroll.okta_verify.switch.channel.link.text',
      },
    } as Record<StepperNavButtonConfigDirection, StepperNavButtonConfigAttrs>,
  };

  if (hasQrCode) {
    set(stepperNavBtnConfig, 'navButtonsConfig.prev', {
      variant: 'secondary',
      label: 'next.enroll.okta_verify.switch.channel.link.text',
    } as StepperNavButtonConfigAttrs);
  }

  return stepperNavBtnConfig;
};

export const getTitleKey = (selectedChannel?: string): string => {
  switch (selectedChannel) {
    case 'email':
      return 'oie.enroll.okta_verify.setup.email.title';
    case 'sms':
      return 'oie.enroll.okta_verify.setup.sms.title';
    default:
      return 'oie.enroll.okta_verify.setup.title';
  }
};
