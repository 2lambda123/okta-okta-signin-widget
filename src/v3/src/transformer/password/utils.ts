/*
 * Copyright (c) 2022-present, Okta, Inc. and/or its affiliates. All rights reserved.
 * The Okta software accompanied by this notice is provided pursuant to the Apache License, Version 2.0 (the "License.")
 *
 * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0.
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *
 * See the License for the specific language governing permissions and limitations under the License.
 */

import { PasswordWithConfirmationElement, UISchemaElement } from '../../types';

export const getPasswordMatchingKey = (
  data: Record<string, unknown>,
): string | undefined => {
  if ('credentials.passcode' in data) {
    return 'credentials.passcode';
  }

  if ('credentials.password' in data) {
    return 'credentials.password';
  }

  if ('credentials.newPassword' in data) {
    return 'credentials.newPassword';
  }

  // Should never hit this case as it should be one of values defined above
  return undefined;
};

export const updatePasswordWithConfirmationBtnLabel = (
  elements: UISchemaElement[],
  label: string,
): void => {
  const pwEnrollmentEle = elements.find(
    (element) => element.type === 'PasswordWithConfirmation',
  ) as PasswordWithConfirmationElement;
  if (pwEnrollmentEle) {
    pwEnrollmentEle.options = {
      ...pwEnrollmentEle.options,
      ctaLabel: label,
    };
  }
};
