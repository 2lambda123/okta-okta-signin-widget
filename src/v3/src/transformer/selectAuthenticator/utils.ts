/*
 * Copyright (c) 2022-present, Okta, Inc. and/or its affiliates. All rights reserved.
 * The Okta software accompanied by this notice is provided pursuant to the Apache License, Version 2.0 (the "License.")
 *
 * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0.
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *
 * See the License for the specific language governing permissions and limitations under the License.
 */

import { Input } from '@okta/okta-auth-js';
import { IdxOption } from '@okta/okta-auth-js/lib/idx/types/idx-js';
import { loc } from 'okta';

import { AUTHENTICATOR_ENROLLMENT_DESCR_KEY_MAP, AUTHENTICATOR_KEY } from '../../constants';
import { AuthenticatorButtonElement } from '../../types';

const getAuthenticatorOption = (
  options: IdxOption[],
  authenticatorKey: string,
): IdxOption | undefined => options?.find(
  ({ relatesTo }) => relatesTo?.key === authenticatorKey,
);

export const getOptionValue = (
  inputs: Input[],
  key: string,
): Input | undefined => inputs?.find(
  ({ name }) => name === key,
);

const getAuthenticatorDataSeVal = (authenticatorKey: string, methodType?: string): string => {
  if (authenticatorKey) {
    const method = methodType ? `-${methodType}` : '';
    return `${authenticatorKey}${method}`;
  }
  return '';
};

const buildOktaVerifyOptions = (
  options: IdxOption[],
  isEnroll?: boolean,
): AuthenticatorButtonElement[] => {
  const ovRemediation = getAuthenticatorOption(options, AUTHENTICATOR_KEY.OV);
  const id = (ovRemediation?.value as Input[])?.find(({ name }) => name === 'id')?.value;
  const methodType = (ovRemediation?.value as Input[])?.find(({ name }) => name === 'methodType');
  if (!methodType || !methodType.options?.length) {
    return [];
  }

  return methodType.options.map((option) => ({
    type: 'AuthenticatorButton',
    label: option.label,
    options: {
      key: AUTHENTICATOR_KEY.OV,
      ctaLabel: isEnroll
        ? loc('oie.enroll.authenticator.button.text', 'login')
        : loc('oie.verify.authenticator.button.text', 'login'),
      description: isEnroll && loc(AUTHENTICATOR_ENROLLMENT_DESCR_KEY_MAP[AUTHENTICATOR_KEY.OV], 'login'),
      actionParams: {
        'authenticator.methodType': option.value,
        'authenticator.id': id,
      },
      dataSe: getAuthenticatorDataSeVal(AUTHENTICATOR_KEY.OV, option.value as string),
    },
  })) as AuthenticatorButtonElement[];
};

const getAuthenticatorDescription = (
  options: IdxOption[],
  authenticatorKey: string,
  isEnroll?: boolean,
): string | undefined => {
  if (!authenticatorKey) {
    return undefined;
  }
  if (isEnroll) {
    if (authenticatorKey === AUTHENTICATOR_KEY.ON_PREM) {
      return 'oie.on_prem.authenticator.description';
    }
    return AUTHENTICATOR_ENROLLMENT_DESCR_KEY_MAP[authenticatorKey];
  }

  if (authenticatorKey === AUTHENTICATOR_KEY.PHONE) {
    return getAuthenticatorOption(
      options,
      AUTHENTICATOR_KEY.PHONE,
    )?.relatesTo?.profile?.phoneNumber as string;
  }
  return undefined;
};

const getOnPremDescriptionParams = (
  options: IdxOption[],
  authenticatorKey: string,
  isEnroll?: boolean,
): string[] | undefined => {
  if (!isEnroll || authenticatorKey !== AUTHENTICATOR_KEY.ON_PREM) {
    return undefined;
  }
  const vendorName = getAuthenticatorOption(
    options,
    AUTHENTICATOR_KEY.ON_PREM,
  )?.relatesTo?.displayName || 'oie.on_prem.authenticator.default.vendorName';
  return [vendorName];
};

const formatAuthenticatorOptions = (
  options: IdxOption[],
  isEnroll?: boolean,
): AuthenticatorButtonElement[] => (
  options.map((option: IdxOption) => {
    const authenticatorKey = option.relatesTo?.key as string;
    const id = getOptionValue(option.value as Input[], 'id')?.value;
    const methodType = getOptionValue(option.value as Input[], 'methodType')?.value;
    const enrollmentId = getOptionValue(option.value as Input[], 'enrollmentId')?.value;

    const descr = getAuthenticatorDescription(
      options,
      authenticatorKey,
      isEnroll,
    );
    const descrParams = getOnPremDescriptionParams(
      options,
      authenticatorKey,
      isEnroll,
    );
    return {
      type: 'AuthenticatorButton',
      label: option.label,
      options: {
        key: authenticatorKey,
        ctaLabel: isEnroll
          ? loc('oie.enroll.authenticator.button.text', 'login')
          : loc('oie.verify.authenticator.button.text', 'login'),
        description: loc(descr, descrParams),
        actionParams: {
          'authenticator.id': id,
          'authenticator.methodType': methodType,
          'authenticator.enrollmentId': enrollmentId,
        },
        dataSe: getAuthenticatorDataSeVal(
          authenticatorKey,
          typeof methodType === 'string' ? methodType : undefined,
        ),
      },
    } as AuthenticatorButtonElement;
  })
);

const getAuthenticatorButtonElements = (
  options: IdxOption[],
  isEnroll?: boolean,
): AuthenticatorButtonElement[] => {
  const formattedOptions = formatAuthenticatorOptions(options, isEnroll);

  // appending OV options back to its original spot
  const ovOptions = buildOktaVerifyOptions(options, isEnroll);
  if (ovOptions.length && options?.length) {
    const ovIndex = options.findIndex(({ relatesTo }) => relatesTo?.key === AUTHENTICATOR_KEY.OV);
    formattedOptions.splice(ovIndex, 1, ...ovOptions);
  }

  return formattedOptions;
};

export const getOVMethodTypeAuthenticatorButtonElements = (
  options: IdxOption[],
): AuthenticatorButtonElement[] => {
  if (!options.length) {
    return [];
  }

  return options.map((option) => ({
    type: 'AuthenticatorButton',
    label: option.label,
    options: {
      key: AUTHENTICATOR_KEY.OV,
      ctaLabel: loc('oie.verify.authenticator.button.text', 'login'),
      actionParams: {
        'authenticator.methodType': (option.value as string),
      },
    },
  })) as AuthenticatorButtonElement[];
};

export const isOnlyPushWithAutoChallenge = (
  inputs?: Input[],
): boolean => {
  const methodType = inputs?.find(({ name }) => name === 'methodType');
  const autoChallenge = inputs?.find(({ name }) => name === 'autoChallenge');

  return typeof autoChallenge !== 'undefined'
    && methodType?.options?.length === 1
    && methodType.options[0].value === 'push';
};

export const getAuthenticatorVerifyButtonElements = (
  authenticatorOptions: IdxOption[],
):AuthenticatorButtonElement[] => getAuthenticatorButtonElements(
  authenticatorOptions,
);

export const getAuthenticatorEnrollButtonElements = (
  authenticatorOptions: IdxOption[],
): AuthenticatorButtonElement[] => getAuthenticatorButtonElements(
  authenticatorOptions,
  true,
);
