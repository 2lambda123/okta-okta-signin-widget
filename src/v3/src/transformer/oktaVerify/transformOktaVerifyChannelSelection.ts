/*
 * Copyright (c) 2022-present, Okta, Inc. and/or its affiliates. All rights reserved.
 * The Okta software accompanied by this notice is provided pursuant to the Apache License, Version 2.0 (the "License.")
 *
 * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0.
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *
 * See the License for the specific language governing permissions and limitations under the License.
 */

import { loc } from 'okta';
import { IDX_STEP } from '../../constants';
import {
  ButtonElement,
  ButtonType,
  FieldElement,
  IdxStepTransformer,
  TitleElement,
  UISchemaElement,
} from '../../types';
import { isAndroidOrIOS } from '../../util';
import { getUIElementWithName } from '../utils';

export const CHANNEL_SCOPE = '#/properties/authenticator/properties/channel';
const CHANNEL_TO_LABEL_KEY_MAP: { [channel: string]: string } = {
  qrcode: 'oie.enroll.okta_verify.select.channel.qrcode.label',
  email: 'oie.enroll.okta_verify.select.channel.email.label',
  sms: 'oie.enroll.okta_verify.select.channel.sms.label',
};

export const transformOktaVerifyChannelSelection: IdxStepTransformer = ({
  prevTransaction,
  formBag,
}) => {
  const { data, uischema } = formBag;

  const isAndroidOrIOSView = isAndroidOrIOS();
  // eslint-disable-next-line max-len
  const lastSelectedChannel = prevTransaction?.context.currentAuthenticator.value.contextualData?.selectedChannel as string;
  const elements: UISchemaElement[] = [];

  const titleElement: TitleElement = {
    type: 'Title',
    options: {
      content: isAndroidOrIOSView
        ? loc('oie.enroll.okta_verify.setup.title', 'login')
        : loc('oie.enroll.okta_verify.select.channel.title', 'login'),
    },
  };
  elements.push(titleElement);

  const channelSelectionElement = getUIElementWithName(
    'authenticator.channel',
    formBag.uischema.elements as UISchemaElement[],
  ) as FieldElement;
  channelSelectionElement.label = loc('oie.enroll.okta_verify.select.channel.description', 'login');
  channelSelectionElement.options.format = 'radio';
  const { options: { inputMeta: { options = [] } } } = channelSelectionElement;
  channelSelectionElement.options.customOptions = options
    .filter((opt) => opt.value !== lastSelectedChannel)
    .map((opt) => ({
      value: opt.value as string,
      label: loc(CHANNEL_TO_LABEL_KEY_MAP[opt.value as string], 'login'),
    }));
  elements.push(channelSelectionElement);

  data['authenticator.channel'] = channelSelectionElement.options.customOptions[0].value;

  const submitButton: ButtonElement = {
    type: 'Button',
    label: loc('oform.next', 'login'),
    scope: `#/properties/${ButtonType.SUBMIT}`,
    options: {
      type: ButtonType.SUBMIT,
      step: IDX_STEP.SELECT_ENROLLMENT_CHANNEL,
    },
  };
  elements.push(submitButton);

  if (!['email', 'sms'].includes(lastSelectedChannel)) {
    const switchChannelButton = {
      type: 'Button',
      label: loc('next.enroll.okta_verify.switch.channel.link.text', 'login'),
      options: {
        type: ButtonType.BUTTON,
        step: 'select-enrollment-channel',
        variant: 'secondary',
      },
    };
    elements.push(switchChannelButton);
  }

  uischema.elements = elements;

  return formBag;
};
