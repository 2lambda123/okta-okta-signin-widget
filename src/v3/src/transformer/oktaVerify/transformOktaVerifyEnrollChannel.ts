/*
 * Copyright (c) 2022-present, Okta, Inc. and/or its affiliates. All rights reserved.
 * The Okta software accompanied by this notice is provided pursuant to the Apache License, Version 2.0 (the "License.")
 *
 * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0.
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *
 * See the License for the specific language governing permissions and limitations under the License.
 */

import {
  ButtonElement,
  ButtonType,
  DescriptionElement,
  IdxStepTransformer,
  TitleElement,
  UISchemaElement,
} from '../../types';
import { getUIElementWithName } from '../utils';

const CHANNELS = ['phoneNumber', 'email'];
const CHANNEL_TO_KEY_MAP: {
  title: { [channel: string]: string },
  description: { [channel: string]: string },
} = {
  description: {
    email: 'oie.enroll.okta_verify.channel.email.description',
    sms: 'oie.enroll.okta_verify.channel.sms.description',
  },
  title: {
    email: 'oie.enroll.okta_verify.enroll.channel.email.title',
    sms: 'oie.enroll.okta_verify.enroll.channel.sms.title',
  },
};

export const transformOktaVerifyEnrollChannel: IdxStepTransformer = ({
  transaction,
  formBag,
}) => {
  const { context } = transaction;
  const authenticator = context.currentAuthenticator.value;
  const { uischema } = formBag;
  const selectedChannel = authenticator.contextualData?.selectedChannel;
  if (!selectedChannel) {
    return formBag;
  }

  const elements: UISchemaElement[] = [];

  elements.push({
    type: 'Title',
    options: {
      content: CHANNEL_TO_KEY_MAP.title[selectedChannel],
    },
  } as TitleElement);

  CHANNELS.forEach((channelName) => {
    const element = getUIElementWithName(
      channelName,
      uischema.elements as UISchemaElement[],
    );
    if (element) {
      elements.push(element);
    }
  });

  elements.push({
    type: 'Description',
    options: {
      content: CHANNEL_TO_KEY_MAP.description[selectedChannel],
    },
  } as DescriptionElement);

  elements.push({
    type: 'Button',
    label: 'oie.enroll.okta_verify.setupLink',
    scope: '#/properties/setupLink',
    options: {
      type: ButtonType.SUBMIT,
    },
  } as ButtonElement);

  elements.push({
    type: 'Button',
    label: 'next.enroll.okta_verify.switch.channel.link.text',
    options: {
      type: ButtonType.BUTTON,
      variant: 'secondary',
      step: 'select-enrollment-channel',
      stepToRender: 'select-enrollment-channel',
    },
  } as ButtonElement);

  uischema.elements = elements;

  return formBag;
};
