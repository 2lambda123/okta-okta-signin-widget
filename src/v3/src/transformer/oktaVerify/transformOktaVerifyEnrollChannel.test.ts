/*
 * Copyright (c) 2022-present, Okta, Inc. and/or its affiliates. All rights reserved.
 * The Okta software accompanied by this notice is provided pursuant to the Apache License, Version 2.0 (the "License.")
 *
 * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0.
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *
 * See the License for the specific language governing permissions and limitations under the License.
 */

import { IdxContext } from '@okta/okta-auth-js';
import { getStubTransactionWithNextStep } from 'src/mocks/utils/utils';
import {
  ButtonElement,
  DescriptionElement,
  FieldElement,
  FormBag,
  StepperButtonElement,
  StepperLayout,
  TitleElement,
  UISchemaLayout,
  UISchemaLayoutType,
  WidgetProps,
} from 'src/types';

import * as channelTransformer from './transformOktaVerifyChannelSelection';
import { transformOktaVerifyEnrollChannel } from './transformOktaVerifyEnrollChannel';

describe('TransformOktaVerifyEnrollChannel Tests', () => {
  const transaction = getStubTransactionWithNextStep();
  const mockProps: WidgetProps = {};
  let formBag: FormBag;
  let channelSelectFormBag: FormBag;
  let channelTransformerStub: jest.SpyInstance;

  beforeEach(() => {
    formBag = {
      dataSchema: {},
      schema: {},
      uischema: {
        type: UISchemaLayoutType.VERTICAL,
        elements: [],
      },
      data: {},
    };

    channelSelectFormBag = {
      dataSchema: {},
      schema: {},
      uischema: {
        type: UISchemaLayoutType.VERTICAL,
        elements: [{ type: 'Title', options: { content: 'Select a channel' } } as TitleElement],
      },
      data: {},
    };

    channelTransformerStub = jest.spyOn(
      channelTransformer,
      'transformOktaVerifyChannelSelection',
    ).mockReturnValue(channelSelectFormBag);
  });

  afterEach(() => {
    channelTransformerStub.mockReset();
  });

  afterAll(() => {
    jest.restoreAllMocks();
  });

  it('should append email input field to elements list when email is the selectedChannel', () => {
    transaction.context = {
      // TODO: OKTA-503490 temporary sln access missing relatesTo obj
      currentAuthenticator: {
        value: {
          contextualData: { selectedChannel: 'email' },
        },
      },
    } as unknown as IdxContext;
    formBag.uischema.elements.push({ type: 'Control', name: 'email' } as FieldElement);

    const updatedFormBag = transformOktaVerifyEnrollChannel(transaction, formBag, mockProps);

    // Main Stepper
    expect(updatedFormBag.uischema.elements.length).toBe(1);
    // Steps
    expect((updatedFormBag.uischema.elements[0] as StepperLayout).elements.length).toBe(2);
    // First step steps
    expect(
      ((updatedFormBag.uischema.elements[0] as StepperLayout).elements[0] as UISchemaLayout)
        .elements.length,
    ).toBe(5);
    expect(
      (((updatedFormBag.uischema.elements[0] as StepperLayout).elements[0] as UISchemaLayout)
        .elements[0] as TitleElement).options?.content,
    ).toBe('oie.enroll.okta_verify.enroll.channel.email.title');
    expect(
      (((updatedFormBag.uischema.elements[0] as StepperLayout).elements[0] as UISchemaLayout)
        .elements[1] as FieldElement).name,
    ).toBe('email');
    expect(
      (((updatedFormBag.uischema.elements[0] as StepperLayout).elements[0] as UISchemaLayout)
        .elements[2] as DescriptionElement).options?.content,
    ).toBe('oie.enroll.okta_verify.channel.email.description');
    expect(
      (((updatedFormBag.uischema.elements[0] as StepperLayout).elements[0] as UISchemaLayout)
        .elements[3] as ButtonElement).label,
    ).toEqual('oie.enroll.okta_verify.setupLink');
    expect(
      (((updatedFormBag.uischema.elements[0] as StepperLayout).elements[0] as UISchemaLayout)
        .elements[4] as StepperButtonElement).label,
    ).toBe('oie.enroll.okta_verify.switch.channel.link.text');
    expect(
      (((updatedFormBag.uischema.elements[0] as StepperLayout).elements[0] as UISchemaLayout)
        .elements[4] as StepperButtonElement).options?.variant,
    ).toBe('secondary');
    expect(
      (((updatedFormBag.uischema.elements[0] as StepperLayout).elements[0] as UISchemaLayout)
        .elements[4] as StepperButtonElement).options?.nextStepIndex,
    ).toBe(1);

    // Second step steps
    expect(
      ((updatedFormBag.uischema.elements[0] as StepperLayout).elements[1] as UISchemaLayout)
        .elements.length,
    ).toBe(1);
    expect(
      (((updatedFormBag.uischema.elements[0] as StepperLayout).elements[1] as UISchemaLayout)
        .elements[0] as TitleElement).options?.content,
    ).toBe('Select a channel');
  });

  it('should append phone input field to elements list when sms is the selectedChannel', () => {
    transaction.context = {
      // TODO: OKTA-503490 temporary sln access missing relatesTo obj
      currentAuthenticator: {
        value: {
          contextualData: { selectedChannel: 'sms' },
        },
      },
    } as unknown as IdxContext;
    formBag.uischema.elements.push({ type: 'Control', name: 'phoneNumber' } as FieldElement);

    const updatedFormBag = transformOktaVerifyEnrollChannel(transaction, formBag, mockProps);

    // Main Stepper
    expect(updatedFormBag.uischema.elements.length).toBe(1);
    // Steps
    expect((updatedFormBag.uischema.elements[0] as StepperLayout).elements.length).toBe(2);
    // First step steps
    expect(
      ((updatedFormBag.uischema.elements[0] as StepperLayout).elements[0] as UISchemaLayout)
        .elements.length,
    ).toBe(5);
    expect(
      (((updatedFormBag.uischema.elements[0] as StepperLayout).elements[0] as UISchemaLayout)
        .elements[0] as TitleElement).options?.content,
    ).toBe('oie.enroll.okta_verify.enroll.channel.sms.title');
    expect(
      (((updatedFormBag.uischema.elements[0] as StepperLayout).elements[0] as UISchemaLayout)
        .elements[1] as FieldElement).name,
    ).toBe('phoneNumber');
    expect(
      (((updatedFormBag.uischema.elements[0] as StepperLayout).elements[0] as UISchemaLayout)
        .elements[2] as DescriptionElement).options?.content,
    ).toBe('oie.enroll.okta_verify.channel.sms.description');
    expect(
      (((updatedFormBag.uischema.elements[0] as StepperLayout).elements[0] as UISchemaLayout)
        .elements[3] as ButtonElement).label,
    ).toEqual('oie.enroll.okta_verify.setupLink');
    expect(
      (((updatedFormBag.uischema.elements[0] as StepperLayout).elements[0] as UISchemaLayout)
        .elements[4] as StepperButtonElement).label,
    ).toBe('oie.enroll.okta_verify.switch.channel.link.text');
    expect(
      (((updatedFormBag.uischema.elements[0] as StepperLayout).elements[0] as UISchemaLayout)
        .elements[4] as StepperButtonElement).options?.variant,
    ).toBe('secondary');
    expect(
      (((updatedFormBag.uischema.elements[0] as StepperLayout).elements[0] as UISchemaLayout)
        .elements[4] as StepperButtonElement).options?.nextStepIndex,
    ).toBe(1);

    // Second step steps
    expect(
      ((updatedFormBag.uischema.elements[0] as StepperLayout).elements[1] as UISchemaLayout)
        .elements.length,
    ).toBe(1);
    expect(
      (((updatedFormBag.uischema.elements[0] as StepperLayout).elements[1] as UISchemaLayout)
        .elements[0] as TitleElement).options?.content,
    ).toBe('Select a channel');
  });
});
