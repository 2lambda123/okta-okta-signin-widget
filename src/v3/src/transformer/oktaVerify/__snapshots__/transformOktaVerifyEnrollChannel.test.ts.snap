// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TransformOktaVerifyEnrollChannel Tests should append email input field to elements list when email is the selectedChannel 1`] = `
Object {
  "data": Object {},
  "dataSchema": Object {},
  "schema": Object {},
  "uischema": Object {
    "elements": Array [
      Object {
        "options": Object {
          "content": "oie.enroll.okta_verify.enroll.channel.email.title",
        },
        "type": "Title",
      },
      Object {
        "name": "email",
        "type": "Control",
      },
      Object {
        "options": Object {
          "content": "oie.enroll.okta_verify.channel.email.description",
        },
        "type": "Description",
      },
      Object {
        "label": "oie.enroll.okta_verify.setupLink",
        "options": Object {
          "step": "",
          "type": "submit",
        },
        "scope": "#/properties/setupLink",
        "type": "Button",
      },
      Object {
        "label": "oie.enroll.okta_verify.switch.channel.link.text",
        "options": Object {
          "step": "select-enrollment-channel",
          "stepToRender": "select-enrollment-channel",
          "type": "button",
          "variant": "secondary",
        },
        "type": "Button",
      },
    ],
    "type": "VerticalLayout",
  },
}
`;

exports[`TransformOktaVerifyEnrollChannel Tests should append phone input field to elements list when sms is the selectedChannel 1`] = `
Object {
  "data": Object {},
  "dataSchema": Object {},
  "schema": Object {},
  "uischema": Object {
    "elements": Array [
      Object {
        "options": Object {
          "content": "oie.enroll.okta_verify.enroll.channel.sms.title",
        },
        "type": "Title",
      },
      Object {
        "name": "phoneNumber",
        "type": "Control",
      },
      Object {
        "options": Object {
          "content": "oie.enroll.okta_verify.channel.sms.description",
        },
        "type": "Description",
      },
      Object {
        "label": "oie.enroll.okta_verify.setupLink",
        "options": Object {
          "step": "",
          "type": "submit",
        },
        "scope": "#/properties/setupLink",
        "type": "Button",
      },
      Object {
        "label": "oie.enroll.okta_verify.switch.channel.link.text",
        "options": Object {
          "step": "select-enrollment-channel",
          "stepToRender": "select-enrollment-channel",
          "type": "button",
          "variant": "secondary",
        },
        "type": "Button",
      },
    ],
    "type": "VerticalLayout",
  },
}
`;
